pkg/vsphere/guestinfo/encode.go

```diff
@@ -0,0 +1,291 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package guestinfo
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+	"strconv"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+const (
+	// DefaultTagName value
+	DefaultTagName = "vic"
+	// DefaultPrefix value
+	DefaultPrefix = ""
+)
+
+const (
+	// Invalid value
+	Invalid = 1 << iota
+	// Hidden value
+	Hidden
+	// ReadOnly value
+	ReadOnly
+	// ReadWrite value
+	ReadWrite
+	// NonPersistent value
+	NonPersistent
+	// Volatile value
+	Volatile
+)
+
+// ToString converts a basic type to string
+func ToString(field reflect.Value) string {
+	switch field.Kind() {
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return strconv.FormatInt(field.Int(), 10)
+	case reflect.Bool:
+		return strconv.FormatBool(field.Bool())
+	case reflect.String:
+		return field.String()
+	case reflect.Float32, reflect.Float64:
+		return strconv.FormatFloat(field.Float(), 'E', -1, 64)
+	default:
+		return ""
+	}
+}
+
+// Scope returns the uint representation of scope tag
+func Scope(scopes []string) uint {
+	var scope uint
+	for _, v := range scopes {
+		switch v {
+		case "hidden":
+			scope |= Hidden
+		case "read-only":
+			scope |= ReadOnly
+		case "read-write":
+			scope |= ReadWrite
+		case "non-persistent":
+			scope |= NonPersistent
+		case "volatile":
+			scope |= Volatile
+		default:
+			return Invalid
+		}
+	}
+	return scope
+}
+
+// CalculateKey creates a new key based on the scope and prefix
+func CalculateKey(scopes []string, prefix string, key string) string {
+	scope := Scope(scopes)
+	if scope&Invalid != 0 || scope&NonPersistent != 0 {
+		return ""
+	}
+
+	var guestinfo string
+	var seperator string
+
+	// Trim the whitespaces
+	key = strings.TrimSpace(key)
+
+	if scope&Hidden != 0 {
+		guestinfo = ""
+		seperator = "~"
```

Comment:
I suggest using / here as well - it's a fail safe in terms of read-only and I don't know what the underlying guestinfo mechanism does if you don't use / or . separators.